### PR TITLE
initial file uploads implementation according to #77

### LINKED ISSUE #77: HTTP file uploads specs
To get quickly with file uploads we will create a specific procedure for this as follows
- rmb need to start with file upload enabled, it's an opt in feature so RMBs can choose not to receive file uploads
- all RMBs can process upload command (to upload a file, not to receive it)
- A special command `msgbus.system.file.upload` is handled in local rmb (the uploader) the body of this command must contain the following payload body (the payload part of the message)
  -  `path` to the local file
  - `cmd` the remote service queue to accept the file
- Note an `upload` command must have a single destination. can't fan out.
- RMB will then start doing an http multipart upload to the remote /rmb-upload/<id> 
  - The request must contain an extra headers:
    - rmb-upload-cmd (this is the cmd part of the request above) this will be notified once upload is complete
    - rmb-source-id (the source twin id)
    - rmb-timestamp (the timestamp of the request)
    - rmb-signature (the signature of the combined headers + id from the url) 
- In case /rmb-upload is disabled, this will return a Method Not Allowed 405 error.
- If upload is allowed:
  - rmb validate headers and signatures. then proceed to writing the file to a temp location (configure on start)
- Once upload is complete remote rmb will issue a command to local service on the cmd queue specified by `rmb-upload-cmd` 
- On local rmb, once uploading is complete, a local message is created on the return queue on the original msgbus.system.file.upload message with the "status" of the upload (this can be success or error)
- There is no grantee that after the upload is successful that the remote service will actually process the uploaded file.

### PR DESCRIPTION
**Pull request by abom**
Closes #77.

TODO:

- [x] Send success message as a reply in local rmb (failure is handled) and send a local message command in remote rmb

Can be done later:

* HttpApi needs a builder because it takes too many parameters now.

### CODE REVIEW

**Comment by abom** [comment]
To be fixed:

* The file should have a unique name (currently it can be overwritten by any new upload with the same file name).

File: src/http_workers/work_runner.rs [hunk 1]
```diff
@@ -174,10 +224,19 @@ where
             // signing the message
             msg.sign(&self.identity);
 
-            result = self.send_once(&twin, &queue, msg).await;
+            if *queue == Queue::Upload {
```
Comment (muhamadazmy) [inline src/http_workers/work_runner.rs:227]: Use a match patter instead of `if`

File: src/http_workers/work_runner.rs [hunk 2]
```diff
@@ -94,11 +97,58 @@ where
...
+            .header(
+                header::CONTENT_TYPE,
+                format!("multipart/form-data; boundary={}", mpart.get_boundary()),
+            )
+            .header("rmb-upload-cmd", payload.cmd)
+            .header("rmb-source-id", payload.source)
+            .header("rmb-timestamp", payload.timestamp)
+            .header(
+                "rmb-signature",
+                payload.signature.unwrap_or_else(|| "".to_string()),
```
Comment (muhamadazmy) [inline src/http_workers/work_runner.rs:123]: I think this should be an error. if signature is not set. It need to be an error. 
I know that internally the get_upload_payload will fail if it failed to sign it but this code makes it unclear and also shows that if signature is not set it's possible to send an empty signature. which is invalid.

Instead, 
```rust
let payload = <get the payload someone> // check other comment on proper way to construct this
let signature = payload.sign()?; // that's one way to do it
```

A more abstract way to do it is to construct MultiPart request from the Payload. In general this code should look like

```rust
let mpart = payload.to_request(&identity)?; 
```
this then can implement and hide/abstract:
- what headers are set 
- how request is signed,
- etc ...

File: src/http_workers/work_runner.rs [hunk 3]
```diff
@@ -94,11 +97,58 @@ where
...
+            )
+            .body(Body::wrap_stream(mpart))
+            .map_err(|err| anyhow!("{:?}", err))?;
+
+        let response = self
+            .client
+            .request(request)
+            .await
+            .map_err(|err| anyhow!("{:?}", err))?;
```
Comment (muhamadazmy) [inline src/http_workers/work_runner.rs:132]: This map (and the one at line 126) will implicitly converted to a SendError::Error which means the upload will be retried. Is this a desired behavior ? 

I prefer instead an explicit conversion.
```suggestion
            .await
            .map_err(|err| SendError::Error(err.into()))?;
```

File: src/http_workers/work_runner.rs [hunk 4]
```diff
@@ -94,11 +97,58 @@ where
...
+            )
+            .header("rmb-upload-cmd", payload.cmd)
+            .header("rmb-source-id", payload.source)
+            .header("rmb-timestamp", payload.timestamp)
+            .header(
+                "rmb-signature",
+                payload.signature.unwrap_or_else(|| "".to_string()),
+            )
+            .body(Body::wrap_stream(mpart))
+            .map_err(|err| anyhow!("{:?}", err))?;
```
Comment (muhamadazmy) [inline src/http_workers/work_runner.rs:126]: If you failed to create the request object, that is definitely a Terminal error hence
```suggestion
            .map_err(|err| SendError::Terminal(err.into()))?;
```

**Review by muhamadazmy — CHANGES_REQUESTED** [review]
Good job. this is definitely up to specs. Please check my comments regarding some code improvements.

File: src/http_api/entry.rs [hunk 5]
```diff
@@ -223,7 +231,162 @@ pub async fn rmb_reply<S: Storage, I: Identity, D: TwinDB>(
...
+    Ok(stream)
+}
+
+async fn process_multipart_field<'a, S: Storage, I: Identity, D: TwinDB>(
+    data: &AppData<S, I, D>,
+    payload: &mut UploadPayload,
+    field: &mut MultipartField<&'a mut Body, hyper::Error>,
+) -> Result<()> {
+    log::debug!("Field received:{}", field.name().unwrap_or_default());
+    if let Ok(filename) = field.filename() {
```
[the remark this answers is not in the document]
Comment (abom) [reply src/http_api/entry.rs:319]: Yes, the overriding of files is avoided by generating a unique directory to write this file into and the full path is sent to the service.

File: src/http_api/entry.rs [hunk 6]
```diff
@@ -223,7 +231,162 @@ pub async fn rmb_reply<S: Storage, I: Identity, D: TwinDB>(
...
+        let parent_dir =
+            Path::new(&data.upload_config.files_path).join(format!("{}", uuid::Uuid::new_v4()));
+
+        fs::create_dir_all(&parent_dir).with_context(|| "cannot create the parent directory")?;
+
+        let path_buf = parent_dir.join(&filename);
+        // to make it consistent between here and the processor
+        let path = path_buf.to_string_lossy();
+        payload.path = path.to_string();
+        let mut file = File::create(&payload.path).with_context(|| "cannot create the file")?;
```
[the remark this answers is not in the document]
Comment (abom) [reply src/http_api/entry.rs:333]: I think the reason was because `path` was `Cow<_, str>` or something, will just create a string from `path_buf` directly.

File: src/main.rs [hunk 7]
```diff
@@ -132,6 +144,27 @@ async fn app(args: &Args) -> Result<()> {
         },
     };
 
+    // uploads config
+    let tmp_path = env::temp_dir();
+    let files_path = match args.files_path.as_ref() {
```
[the remark this answers is not in the document]
Comment (abom) [reply src/main.rs:149]: I named it like this not to have a conflict with the short option, will rename it and make only a long one.

File: src/http_workers/work_runner.rs (hunk 3 above)
Comment (abom) [reply src/http_workers/work_runner.rs:132]: For retries yes, just like normal message

File: src/types/mod.rs [hunk 8]
```diff
@@ -93,34 +172,52 @@ impl Message {
...
-        // we do unwrap because this should never fail.
-        let digest = self.challenge().unwrap();
-        let signature = signer.sign(&digest[..]);
+        self.signature = Some(sign(self, signer));
+    }
+
+    /// verify the message signature
+    pub fn verify<I: Identity>(&self, identity: I) -> Result<()> {
+        verify(self, &identity, &self.signature)
+    }
```
[the remark this answers is not in the document]
Comment (abom) [reply src/types/mod.rs]: I don't remember, got it back :smile:

File: src/proxy/mod.rs [hunk 9]
```diff
@@ -62,7 +62,7 @@ where
 
         // verify before modifying the message
         payload
-            .verify(&twin.account)
+            .verify(twin.account)
```
[the remark this answers is not in the document]
Comment (abom) [reply src/proxy/mod.rs:65]: I don't remember, got it back :smile:

File: src/types/mod.rs [hunk 10]
```diff
@@ -4,12 +4,31 @@ use anyhow::{Context, Result};
...
+    #[serde(default)]
+    #[serde(skip_deserializing)]
+    pub timestamp: u64,
+
+    #[serde(default)]
+    #[serde(skip_deserializing)]
+    pub source: u32,
+
+    #[serde(skip_deserializing)]
+    pub signature: Option<String>,
```
[the remark this answers is not in the document]
Comment (muhamadazmy) [reply src/types/mod.rs:31]: I am sorry if this is a lot of work but it will make things much cleaner because now it's really hard to follow the cycle of signing/verification since fields are available in all context while they only be visible in a subset of context.

In other words. path and cmd all available in client, local rmb, remote rmb, server. while signature source, and timestamp are only available in local rmb, and remote rmb. Hence i can't just have all fields everywhere.

If the client can't set the signature or the timestamp, then he shouldn't be able to even set them

[the remark this answers is not in the document]
Comment (muhamadazmy) [reply src/types/mod.rs:31]: To make it easy to use the Challengeable interface to calculate the signature you can add this 

```rust
impl<T> Challengeable for &[T]
where
    T: std::fmt::Display,
{
    fn challenge(&self) -> Result<md5::Digest> {
        let mut hash = md5::Context::new();
        for v in self.iter() {
            write!(hash, "{}", v)?;
        }

        Ok(hash.compute())
    }
}
```

then you can simply do 

```rust
let fields = vec![ put all fields here (as strings) ];
sign(&fields);
```

**Review by muhamadazmy — CHANGES_REQUESTED** [review]

File: src/http_api/upload.rs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        Ok(stream)
+    }
+
+    async fn process_multipart_field(
+        &self,
+        payload: &mut UploadPayload,
+        field: &mut MultipartField<&'a mut Body, hyper::Error>,
+    ) -> Result<()> {
+        // if a file is provided, we always generate a uuid for as filename
+        if field.filename().is_ok() {
```
[the remark this answers is not in the document]
Comment (abom) [reply src/http_api/upload.rs:112]: Seems like I got confused with another crate that has different types of fields, will remove this, `filename` is not needed anyway.

File: src/http_api/errors.rs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+use thiserror::Error;
```
[the remark this answers is not in the document]
Comment (abom) [reply src/http_api/errors.rs:1]: Ah, I didn't see the `impl` part. I moved `HandlerError` to to a common module because two other modules need to use it, and for sure we can start re-organizing `http_api` module.

File: src/http_api/upload.rs [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+            Err(err) => return Err(HandlerError::BadRequest(err)),
+        };
+
+        let mut stream = match self.get_multipart_stream(&mut request) {
+            Ok(s) => s,
+            Err(err) => return Err(HandlerError::BadRequest(err)),
+        };
+
+        while let Ok(Some(mut field)) = stream.try_next().await {
+            if let Err(err) = self.process_multipart_field(&mut payload, &mut field).await {
```
[the remark this answers is not in the document]
Comment (abom) [reply src/http_api/upload.rs:145]: It would complain because of the loop?, I'll pass only the reference and won't set the path in the payload itself.

**Review by muhamadazmy — DISMISSED** [review]
